Use conda + Update initial setup docs (#1163)
ucb-bar/chipyard: docs/Chipyard-Basics/Initial-Repo-Setup.rst

Afterwards, verify that Conda is a sufficient version (we test on version 4.12.0 but higher is most likely fine).

.. Note:: If you have installed conda separately from this documentation (i.e. from miniconda or full Anaconda), please ensure you follow https://conda-forge.org/docs/user/introduction.html#how-can-i-install-packages-from-conda-forge to use ``conda-forge`` packages without any issues.

.. code-block:: shell

    conda --version # must be version 4.12.0 or higher

After Conda is installed and is on your ``PATH``, we need to install a version of ``git`` to initially checkout the repository.
For this you can use the system package manager like ``yum`` or ``apt`` to install ``git``.
This ``git`` is only used to first checkout the repository, we will later install a newer version of ``git`` with Conda.

Finally we need to install ``conda-lock`` into the ``base`` conda environment.
This is done by the following:

.. code-block:: shell

    conda install -n base conda-lock
    conda activate base

Setting up the Chipyard Repo
-------------------------------------------

> Start by checking out the proper Chipyard's version. Run:

Reviewer: ```suggestion
Start by checking out the proper Chipyard version. Run:
```

Chipyard is developed and tested on Linux-based systems.

.. Warning:: It is possible to use this on macOS or other BSD-based systems, although GNU tools will need to be installed;
    it is also recommended to install the RISC-V toolchain from ``brew``.

.. Warning:: Working under Windows is not recommended.

Running on AWS EC2 with FireSim
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If you plan on using Chipyard alongside FireSim, you should refer to the :fsim_doc:`FireSim documentation <>`.
Specifically, you should follow the :fsim_doc:`Initial Setup/Simulation <Initial-Setup/index.html>`
section of the docs up until :fsim_doc:`Setting up the FireSim Repo <Initial-Setup/Setting-up-your-Manager-Instance.html#setting-up-the-firesim-repo>`.
At that point, instead of checking out FireSim you can checkout Chipyard by following :ref:`Chipyard-Basics/Initial-Repo-Setup:Setting up the Chipyard Repo`.

Default Requirements Installation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In Chipyard, we use the `Conda <https://docs.conda.io/en/latest/>`__ package manager to help manage system dependencies.
Conda allows users to create an "environment" that holds system dependencies like ``make``, ``gcc``, etc.

> 

Reviewer 2: ```suggestion

.. Note:: Chipyard can also run on systems without a Conda installation. However, users on these systems must manually install toolchains and dependencies.

```

First Chipyard requires there to be Conda installed on the system.

> Please refer to the `Conda installation instructions <https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html>`__ on how to install Conda.

Reviewer 3: If people install conda from one of the Anaconda Inc installers (miniconda or the full Anaconda), you will be configured to install packages from the 'defaults' channels built into conda (that come from Anaconda Inc).

'defaults' and 'conda-forge' are known to conflict in some gnarly ways and that's why `conda-forge` folks reccomend settings documented at https://conda-forge.org/docs/user/introduction.html#how-can-i-install-packages-from-conda-forge 

For more information on the types of conflicts that happen, see https://conda-forge.org/docs/user/tipsandtricks.html#multiple-channels

Alternately, you can avoid the whole issue by installing from one of the miniforge installers https://github.com/conda-forge/miniforge/#download which don't use anything built by Anaconda Inc and don't have the 'defaults' channel available.

Since you're using `conda env create -f scripts/conda-requirements.yml`, I _think_ the `channels` list in the YAML overrides any channels configured.  I'll go double check that I'm remembering that correctly.

Author: Interesting, I'll try to update the docs and point to the miniforge installers since that's probably the best thing to do.

Afterwards, verify that Conda is a sufficient version (we test on version 4.12.0).

.. code-block:: shell

    conda --version # must be version 4.12.0 or higher

After Conda is installed and is on your ``PATH``, we need to install a version of ``git`` to initially checkout the repository.
For this you can use the system package manager like ``yum`` or ``apt`` to install ``git``.
This ``git`` is only used to first checkout the repository, we will later install a newer version of ``git`` with Conda.

Setting up the Chipyard Repo
-------------------------------------------

> Start by checkout out the proper Chipyard's version. Run:

Reviewer 2: ```suggestion
Start by checking out out the proper Chipyard's version. Run:
```

Chipyard is developed and tested on Linux-based systems.

.. Warning:: It is possible to use this on macOS or other BSD-based systems, although GNU tools will need to be installed;
    it is also recommended to install the RISC-V toolchain from ``brew``.

.. Warning:: Working under Windows is not recommended.

Running on AWS EC2 with FireSim
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If you plan on using Chipyard alongside FireSim, you should refer to the :fsim_doc:`FireSim documentation <>`.
Specifically, you should follow the :fsim_doc:`Initial Setup/Simulation <Initial-Setup/index.html>`
section of the docs up until :fsim_doc:`Setting up the FireSim Repo <Initial-Setup/Setting-up-your-Manager-Instance.html#setting-up-the-firesim-repo>`.

> At that point, instead of checking out FireSim you can checkout Chipyard by following :ref:`Chipyard-Basics/Initial-Repo-Setup:Setting up the Chipyard Repo`.

Reviewer: ```suggestion
At that point, instead of cloning FireSim you can clone Chipyard by following :ref:`Chipyard-Basics/Initial-Repo-Setup:Setting up the Chipyard Repo`.
```

Default Requirements Installation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In Chipyard, we use the `Conda <https://docs.conda.io/en/latest/>`__ package manager to help manage system dependencies.
Conda allows users to create an "environment" that holds system dependencies like ``make``, ``gcc``, etc.

.. Note:: Chipyard can also run on systems without a Conda installation. However, users on these systems must manually install toolchains and dependencies.

> First Chipyard requires there to be Conda installed on the system.

Reviewer: ```suggestion
First, Chipyard requires Conda to be installed on the system.
```

.. Warning:: Working under Windows is not recommended.

Running on AWS EC2 with FireSim
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If you plan on using Chipyard alongside FireSim, you should refer to the `FireSim documentation `<https://docs.fires.im/>`__
on "Initial Setup/Simulation".
When you reach Section 2.3.2 on "Setting up the FireSim Repo"

> you can continue to :ref:`Setting up the Chipyard Repo` to continue to setup the repository with Chipyard as the top-level repository.

Reviewer: Can you directly link to the correct starting section of the FireSim docs?

Author: That's the plan. I just need to figure out to point to either latest or stable depending on what the version of the CY docs is being built.

Default Requirements Installation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In Chipyard, we use `Conda <https://docs.conda.io/en/latest/>`__ to help manage system dependencies (``make``, ``gcc``, etc).

> We **highly** recommend running the following script to setup Chipyard's Conda environment.

Reviewer 2: Can you add a description here of what the conda env is pulling?

.. code-block:: shell

    ./scripts/machine-launch-script.sh

> .. Warning:: If you have ``sudo`` priveleges then this script should work fine.

Reviewer 2: ```suggestion
.. Warning:: If you have ``sudo`` privileges then this script should work fine.
```

    However, if you are not a system administrator (i.e. you are on a shared system), you
    should run ``./scripts/machine-launch-script.sh --help`` to view relevant configuration parameters.

.. Note:: When running on an Amazon Web Services EC2 FPGA-development instance
    (for FireSim), FireSim includes a similar machine setup script that will install all
    of the aforementioned dependencies (and some additional ones).

Depending on where you ran the machine launch script,
you should find a file called ``machine-launch-status.log`` and ``machine-launch-status`` indicating
if things completed properly.

> 

Reviewer 2: Why does this log it to a status file, instead of just showing it on the terminal.

Author: It does both. The docs can be clarified

.. code-block:: shell

    cat ./machine-launch-status # should have "started" and "completed"

> Once run, please confirm that you "activated" the proper Conda environment and the prior script completed.

Reviewer 2: Should there be instructions on how to activate a Conda env?

By running the following command you should see "chipyard" output (the default environment is called "chipyard").

.. code-block:: shell

    echo $CONDA_DEFAULT_ENV

.. Note:: If you don't have a Conda environment set (and the machine launch script completed),

>     you can run ``conda activate chipyard`` to "enter" the proper environment.

Reviewer 2: Does the launch script auto activate the env? 
I'm not sure I like the idea of scripts which hijack the environment of the user.

Conda Environment Setup
~~~~~~~~~~~~~~~~~~~~~~~

.. Warning:: When running on an Amazon Web Services EC2 FPGA-development instance
    (for FireSim), FireSim includes a similar machine setup script that will install all
    of the aforementioned dependencies (and some additional ones) and will activate the
    proper conda environment. Skip this section.

Next run the following command to create Chipyard's Conda environment.

.. code-block:: shell

    conda env create -f scripts/conda-requirements.yml

By running the following command you should see a "chipyard" environment listed (the default environment is called "chipyard" and can be modified with ``conda`` arguments).

.. code-block:: shell

    conda env list

.. Note:: Refer to FireSim's :fsim_doc:`Conda documentation <Advanced-Usage/Conda.html>` on more information
    on how to use Conda and some of its benefits.

Next go ahead and activate the conda environment that was setup.

.. code-block:: shell

    conda activate chipyard

> We recommend that you add this "activate" command your ``.bashrc`` (or other environment setup file).

Reviewer 2: ```suggestion
We recommend that you add this "activate" command to your ``.bashrc`` (or other environment setup file).
```

Fetch Chipyard Sources
~~~~~~~~~~~~~~~~~~~~~~

To fetch all Chipyard sources, run the following:

.. code-block:: shell

    ./scripts/init-submodules-no-riscv-tools.sh

This will initialize and checkout all of the necessary git submodules.
This will also validate that you are on a tagged branch, otherwise it will prompt for confirmation.

When updating Chipyard to a new version, you will also want to rerun this script to update the submodules.
Using ``git`` directly will try to initialize all submodules; this is not recommended unless you expressly desire this behavior.

Obtaining a Toolchain
------------------------

Currently there are two toolchains, one for normal RISC-V programs called ``riscv-tools``, and another for Hwacha/Gemmini called ``esp-tools``.
To get a basic ``riscv-tools`` installation (which is the only thing needed for most Chipyard use-cases), just the following steps are necessary.

.. code-block:: shell

    conda install -c ucb-bar riscv-tools # for a normal risc-v toolchain

> .. Note:: If you are planning to use the Hwacha vector unit, or other RoCC-based accelerators, you should obtain the ``esp-tools`` toolchain by adding the ``esp-tools`` argument to the command above.

Reviewer 2: Shouldn't this be "substituting" and not "adding" the argument?

Once the command is run, the ``PATH``, ``RISCV``, and ``LD_LIBRARY_PATH`` environment variables will be set properly.

> .. Note:: If you are a power user and would like to build your own toolchain, you can refer to the https://github.com/ucb-bar/riscv-tools-feedstock and https://github.com/ucb-bar/esp-tools-feedstock repositories on how to build a toolchain yourself.

Reviewer 2: Didn't we flatten riscv-tools and esp-tools into chipyard in the first place to make it easier to bump those projects? This adds back that level of indirection.

In fact, doesn't this add two levels of indirection? Now we additionally have to track commits for riscv-tools-feedstock and riscv-tools.

Author: Yes. But for the most part, the main item that is being rebuilt is Spike (for Gemmini folks). That can be rebuilt separately. In the future we will focus on splitting up things even more.

> .. Note:: If can deactivate/activate a toolchain (but keep it installed) by running ``source $CONDA_PREFIX/etc/conda/deactivate.d/deactivate-${PKG_NAME}.sh`` or ``$CONDA_PREFIX/etc/conda/activate.d/activate-${PKG_NAME}.sh`` (``PKG_NAME`` for example is ``ucb-bar-riscv-tools``). This will modify the aforementioned 3 environment variables.

Reviewer 2: ```suggestion
.. Note:: You can deactivate/activate a toolchain (but keep it installed) by running ``source $CONDA_PREFIX/etc/conda/deactivate.d/deactivate-${PKG_NAME}.sh`` or ``$CONDA_PREFIX/etc/conda/activate.d/activate-${PKG_NAME}.sh`` (``PKG_NAME`` for example is ``ucb-bar-riscv-tools``). This will modify the aforementioned 3 environment variables.
```

    # checkout latest official chipyard release
    # note: this may not be the latest release if the documentation version != "stable"
    git checkout |version|

If you are running Chipyard alongside FireSim on AWS EC2, you should skip the :ref:`Chipyard-Basics/Initial-Repo-Setup:Conda Environment Setup` section and instead jump to :ref:`Chipyard-Basics/Initial-Repo-Setup:Fetch Chipyard Sources`.

Conda Environment Setup
~~~~~~~~~~~~~~~~~~~~~~~

.. Warning:: When running on an Amazon Web Services EC2 FPGA-development instance
    (for FireSim), FireSim includes a similar machine setup script that will install all
    of the aforementioned dependencies (and some additional ones) and will activate the
    proper conda environment. Skip this section.

Next run the following command to create Chipyard's Conda environment.

.. code-block:: shell

>     conda env create -f scripts/conda-requirements.yml

Reviewer 4: ```suggestion
    conda env create -f scripts/conda-requirements.yaml -n chipyard
```

The file extension should be `.yaml`.
In my case, the `-n` option turned out to be mandatory.
